src/SSymmetricCompact.jl

```diff
@@ -0,0 +1,269 @@
+
+An `SSymmetricCompact` may be constructed either:
+
+* from an `AbstractVector` containing the lower triangular elements; or
+* from a `Tuple` containing both upper and lower triangular elements (in column major order); or
+* from another `StaticMatrix`.
+
+For the latter two cases, only the lower triangular elements are used; the upper triangular
+elements are ignored.
+"""
+struct SSymmetricCompact{N, T, L} <: StaticMatrix{N, N, T}
+    lowertriangle::SVector{L, T}
+
+    @inline function SSymmetricCompact{N, T, L}(lowertriangle::SVector{L}) where {N, T, L}
+        check_symmetric_parameters(Val(N), Val(L))
+        new{N, T, L}(lowertriangle)
+    end
+end
+
+@inline function check_symmetric_parameters(::Val{N}, ::Val{L}) where {N, L}
+    if 2 * L !== N * (N + 1)
+        throw(ArgumentError("Size mismatch in SSymmetricCompact parameters. Got dimension $N and length $L."))
+    end
+end
+
+Base.@pure triangularnumber(N::Int) = div(N * (N + 1), 2)
+Base.@pure triangularroot(L::Int) = div(isqrt(8 * L + 1) - 1, 2) # from quadratic formula
+
+lowertriangletype(::Type{SSymmetricCompact{N, T, L}}) where {N, T, L} = SVector{L, T}
+lowertriangletype(::Type{SSymmetricCompact{N, T}}) where {N, T} = SVector{triangularnumber(N), T}
+lowertriangletype(::Type{SSymmetricCompact{N}}) where {N} = SVector{triangularnumber(N)}
+
+@inline (::Type{SSymmetricCompact{N, T}})(lowertriangle::SVector{L}) where {N, T, L} = SSymmetricCompact{N, T, L}(lowertriangle)
+@inline (::Type{SSymmetricCompact{N}})(lowertriangle::SVector{L, T}) where {N, T, L} = SSymmetricCompact{N, T, L}(lowertriangle)
+
+@inline function SSymmetricCompact(lowertriangle::SVector{L, T}) where {T, L}
+    N = triangularroot(L)
+    SSymmetricCompact{N, T, L}(lowertriangle)
+end
+
+@generated function (::Type{SSymmetricCompact{N, T, L}})(a::Tuple) where {N, T, L}
+    expr = Vector{Expr}(undef, L)
+    i = 0
+    for col = 1 : N, row = col : N
+        index = N * (col - 1) + row
+        expr[i += 1] = :(a[$index])
+    end
+    quote
+        @_inline_meta
+        @inbounds return SSymmetricCompact{N, T, L}(SVector{L, T}(tuple($(expr...))))
+    end
+end
+
+@inline function (::Type{SSymmetricCompact{N, T}})(a::Tuple) where {N, T}
+    L = triangularnumber(N)
+    SSymmetricCompact{N, T, L}(a)
+end
+
+@inline (::Type{SSymmetricCompact{N}})(a::Tuple) where {N} = SSymmetricCompact{N, promote_tuple_eltype(a)}(a)
+@inline (::Type{SSymmetricCompact{N}})(a::NTuple{M, T}) where {N, T, M} = SSymmetricCompact{N, T}(a)
+@inline SSymmetricCompact(a::StaticMatrix{N, N, T}) where {N, T} = SSymmetricCompact{N, T}(a)
+
+@inline (::Type{SSC})(a::SSymmetricCompact) where {SSC <: SSymmetricCompact} = SSC(a.lowertriangle)
+@inline (::Type{SSC})(a::SSC) where {SSC <: SSymmetricCompact} = a
+
+@inline (::Type{SSC})(a::AbstractVector) where {SSC <: SSymmetricCompact} = SSC(convert(lowertriangletype(SSC), a))
+
+# TODO: similar_type overload?
```

Comment:
I think this might be required, and you'll have to fall back to `SMatrix` for correctness? Unless we introduced a way to handle symmetry-preserving operations.

Reply:
`SMatrix` is the default. I should probably just remove this comment instead.

Reply:
Oh sure, sorry my bad!